app/pages/SearchResultsPage/SearchResultsPage.tsx

```diff
@@ -34,22 +36,80 @@ interface SearchState extends ParsedQs {
 
 /** Wrapper component that handles state management, URL parsing, and external API requests. */
 export const SearchResultsPage = () => {
+  const [cookies] = useCookies(["googtrans"]);
+  const history = useHistory();
+  const { search } = useLocation();
   const { userLocation } = useAppContext();
   const [lastPush, setLastPush] = useState(Date.now());
-  const { search } = useLocation();
   const [expandList, setExpandList] = useState(false);
 
-  const searchState: SearchState = useMemo(
-    () => qs.parse(search.slice(1)),
-    [search]
-  );
+  const [searchState, setSearchState] = useState<SearchState | null>(null);
   const [searchRadius, setSearchRadius] = useState(
     searchState?.configure?.aroundRadius ?? "all"
   );
 
+  // In cases where we translate a query into English, we use this ref
+  // to represent the user's original, untranslated input. The untranslatedQuery
+  // is displayed in the UI and stored in the URL params.
+  // This untranslatedQuery value is also checked against when a new search is triggered
+  // to determine if the user has input a different query (vs. merely selecting refinements),
+  // in which case we need to call the translation API again
+  const untranslatedQuery = useRef<string | null | undefined>(null);
+  const translatedQuery = useRef(null);
+
+  useEffect(() => {
```

Comment:
I wonder if this logic could be simplified if we broke it up into independent `useEffect()` blocks, where we make the control flow similar in both the English and non-English cases. I think there's four phases/data dependencies:

1. URL query parameters -> untranslated query
3. untranslated query + translation language -> translated query (no-op for the `en` case)
4. translated query + other search params -> search state

I think a rough sketch of the separate `useEffect()` calls I was thinking of would look like this:

```tsx
useEffect(() => {
  setUntranslatedQuery(qs.parse(search.slice(1)).query)
  // Maybe we should simultaneously set a second state variable holding
  // just the non-query parts of `search`?
}, [search])

useEffect(() => {
  // Grab query language from cookie
  if (queryLanguage === "en") {
    setTranslatedQuery(untranslatedQuery)
  } else {
    // Do the API call to the translate_text/ endpoint
  }
}, [untranslatedQuery, cookie.gooogtrans, queryLanguage])

useEffect(() => {
  setSearchState({...urlParams, query: translatedQuery}); 
}, [translatedQuery, search /* maybe instead of search, we should depend on the non-query parts of search? */])
```

I'm not sure if my suggestion here is any less code, but I think it makes it a little bit clearer which variables are dependent on other variables, and it also leverage's React's `useEffect()`'s second argument for keeping track of when things need to be updated. I think that would make it less error-prone compared to us basically reimplementing that ourselves in a single `useEffect()` block. I think this would bring this code closer to the [Functional Reactive Programming (FRP) paradigm](https://en.wikipedia.org/wiki/Functional_reactive_programming) that React is loosely based on, which is all about declaring data dependencies between variables and letting the system figure out which ones need to be updated due to dependent variables being updated.

I think the main downside with my suggestion is that this would cause up to four re-renders when updating the search params. However, I think this is mitigated by the

```tsx
  if (searchState === null) {
    return null;
  }
```

since at least for the first page load, the first few rerenders would be cheap. If performance is an issue for unnecessarily rerendering the child component (`InnerSearchResults`), then I think we could make use of [`React.memo()`](https://react.dev/reference/react/memo#reference) (not the same thing as `useMemo()`).

Anyway, this is just a thought I had, and I don't think it's strictly necessary to do it this way. I just thought that this could be a way of structuring this `useEffect()` block differently such that makes the dependencies a little clearer.

Reply:
Great suggestion. I agree with you that breaking it up into multiple useEffect blocks clarifies our code. I've gone ahead and followed your suggestion and it seems to work well. Qualitatively speaking, performance looks good. Thanks for taking the time to lay all of this out.